### ISSUE TITLE
informant: cgroup perma-paused if close to OOM but VM cannot increase

### ISSUE THREAD

**Issue by sharnoff** [neondatabase/autoscaling#44]
## Steps to reproduce:

Edit `vm-deploy.yaml` to reduce maximum CPU and memory slots to 1. Create the VM and run

```
cgexec -g memory:neon-test allocate-loop 512 2560
```

Once it gets close to OOM, `allocate-loop` pauses and appears to make no progress — expected behavior is resuming and getting killed.

**Comment by sharnoff** [comment]
Copying over some digging I [posted in slack](https://neondb.slack.com/archives/C03TN5G758R/p1682410239294789), for future reference.

---

The current technique is the following:
1. Set the cgroup's `memory.high` below the maximum amount that we want to let postgres use
2. When we get a memory event for `memory.high` (i.e. total usage > memory.high):
    1. pause ("freeze") the cgroup
    2. request more memory
    3. unpause ("thaw") after timeout or when more memory is added

Testing was done with a [little C program](https://github.com/neondatabase/autoscaling/blob/40176111c8d6d42ac8407447f9f1d7ca5a54c419/vm-examples/pg14-disk-test/allocate-loop.c) that just slowly allocates a bunch of memory, periodically printing out how much it's allocated so far.

Previously, this strategy appeared to cause the test program to completely stop when `memory.high` was reached and no new memory got added (because the VM was at its limit). However: What's actually going on is that the process is just being very aggressively throttled. Look at the timestamps here:

```
neonvm:~# cgexec -g memory:neon-test allocate-loop 256 2280
Selected: minimum = 256 MiB, maximum = 2280 MiB
mmap created at minimum size
07:53:59.352 :: 256 MiB + 256 KiB, rss = 262912
07:53:59.467 :: 281 MiB + 256 KiB, rss = 288452
07:53:59.581 :: 306 MiB + 256 KiB, rss = 314052
07:53:59.696 :: 331 MiB + 256 KiB, rss = 339652
07:53:59.817 :: 356 MiB + 256 KiB, rss = 365252
07:53:59.933 :: 381 MiB + 256 KiB, rss = 390852
07:54:00.050 :: 406 MiB + 256 KiB, rss = 416452
07:54:00.167 :: 431 MiB + 256 KiB, rss = 442052
07:54:00.288 :: 456 MiB + 256 KiB, rss = 467652
07:54:00.403 :: 481 MiB + 256 KiB, rss = 493252
07:54:00.518 :: 506 MiB + 256 KiB, rss = 518852
07:54:00.632 :: 531 MiB + 256 KiB, rss = 544452
07:54:00.748 :: 556 MiB + 256 KiB, rss = 570052
07:54:00.867 :: 581 MiB + 256 KiB, rss = 595652
07:54:00.981 :: 606 MiB + 256 KiB, rss = 621252
07:54:01.099 :: 631 MiB + 256 KiB, rss = 646852
07:54:01.214 :: 656 MiB + 256 KiB, rss = 672452
07:54:01.329 :: 681 MiB + 256 KiB, rss = 698052
07:54:01.444 :: 706 MiB + 256 KiB, rss = 723652 <--- +120ms
07:54:01.768 :: 731 MiB + 256 KiB, rss = 749284 <--- memory.high hits here
07:54:19.380 :: 756 MiB + 256 KiB, rss = 774932 <--- +18s
07:55:29.057 :: 781 MiB + 256 KiB, rss = 800540 <--- +70s — 500x slower
```

The cgroup's `memory.high` was set to 723 MiB, so this is where we'd expect it to be, but much more severe than I was expecting.

This throttling also may be due to the fact that the process is functionally only allocating memory, so maybe postgres won't get hit quite as badly?

So, to recap: The problem is that when users computes are already at their maximum allotted memory, the current solution may just throttle postgres so that it basically just stops instead of running out of memory.
